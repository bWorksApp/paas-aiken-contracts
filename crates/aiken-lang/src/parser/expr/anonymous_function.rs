use crate::{
    ast,
    expr::{FnStyle, UntypedExpr},
    parser::{annotation, error::ParseError, pattern, token::Token},
};
use chumsky::prelude::*;

pub fn parser(
    sequence: Recursive<'_, Token, UntypedExpr, ParseError>,
) -> impl Parser<Token, UntypedExpr, Error = ParseError> + '_ {
    just(Token::Fn)
        .ignore_then(
            params()
                .separated_by(just(Token::Comma))
                .allow_trailing()
                .delimited_by(just(Token::LeftParen), just(Token::RightParen)),
        )
        .then(just(Token::RArrow).ignore_then(annotation()).or_not())
        .then(sequence.delimited_by(just(Token::LeftBrace), just(Token::RightBrace)))
        .map_with_span(
            |((arguments, return_annotation), body), span| UntypedExpr::Fn {
                arguments,
                body: Box::new(body),
                location: span,
                fn_style: FnStyle::Plain,
                return_annotation,
            },
        )
}

pub fn params() -> impl Parser<Token, ast::UntypedArg, Error = ParseError> {
    // TODO: return a better error when a label is provided `UnexpectedLabel`
    choice((
        select! {Token::DiscardName {name} => name}.map_with_span(|name, span| {
            ast::ArgBy::ByName(ast::ArgName::Discarded {
                label: name.clone(),
                name,
                location: span,
            })
        }),
        select! {Token::Name {name} => name}.map_with_span(|name, span| {
            ast::ArgBy::ByName(ast::ArgName::Named {
                label: name.clone(),
                name,
                location: span,
            })
        }),
        pattern().map(ast::ArgBy::ByPattern),
    ))
    .then(just(Token::Colon).ignore_then(annotation()).or_not())
    .map_with_span(|(by, annotation), span| ast::UntypedArg {
        is_validator_param: false,
        location: span,
        annotation,
        doc: None,
        by,
    })
}

#[cfg(test)]
mod tests {
    use crate::assert_expr;

    #[test]
    fn anonymous_function_basic() {
        assert_expr!(r#"fn (a: Int) -> Int { a + 1 }"#);
    }

    #[test]
    fn anonymous_function_by_pattern_no_annotation() {
        assert_expr!(r#"fn (Foo { my_field }) { my_field * 2 }"#);
    }

    #[test]
    fn anonymous_function_by_pattern_with_annotation() {
        assert_expr!(r#"fn (Foo { my_field } : Foo) { my_field * 2 }"#);
    }

    #[test]
    fn anonymous_function_by_pattern_with_alias() {
        assert_expr!(r#"fn (Foo { my_field, .. } as x) { my_field * my_other_field }"#);
    }
}
