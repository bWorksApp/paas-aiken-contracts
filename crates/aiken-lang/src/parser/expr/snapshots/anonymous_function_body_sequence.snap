---
source: crates/aiken-lang/src/parser/expr/anonymous_function.rs
description: "Code:\n\nfn (a: Int) -> Int {\n    let b = 1\n    a + b\n}"
---
Fn {
    location: 0..46,
    fn_style: Plain,
    arguments: [
        Arg {
            arg_name: Named {
                name: "a",
                label: "a",
                location: 4..5,
                is_validator_param: false,
            },
            location: 4..10,
            annotation: Some(
                Constructor {
                    location: 7..10,
                    module: None,
                    name: "Int",
                    arguments: [],
                },
            ),
            doc: None,
            tipo: (),
        },
    ],
    body: Sequence {
        location: 39..44,
        expressions: [
            Assignment {
                location: 25..34,
                value: UInt {
                    location: 33..34,
                    value: "1",
                    base: Decimal {
                        numeric_underscore: false,
                    },
                },
                patterns: [
                    AssignmentPattern {
                        pattern: Var {
                            location: 29..30,
                            name: "b",
                        },
                        annotation: None,
                        location: 29..30,
                    },
                ],
                kind: Let {
                    backpassing: false,
                },
            },
            BinOp {
                location: 39..44,
                name: AddInt,
                left: Var {
                    location: 39..40,
                    name: "a",
                },
                right: Var {
                    location: 43..44,
                    name: "b",
                },
            },
        ],
    },
    return_annotation: Some(
        Constructor {
            location: 15..18,
            module: None,
            name: "Int",
            arguments: [],
        },
    ),
}
